File: server/vidstream.go
package server

import (
	"fmt"
	"time"

	"github.com/teranos/QNTX/ats/vidstream/vidstream"
)

// handleVidStreamInit initializes the ONNX video engine with model configuration
// Runs asynchronously to avoid blocking the WebSocket message handler
func (c *Client) handleVidStreamInit(msg QueryMessage) {
	c.server.logger.Infow("VidStream init request",
		"client_id", c.id,
		"model_path", msg.ModelPath,
	)

	// Run initialization in background to avoid blocking WebSocket
	go func() {
		c.server.vidstreamMu.Lock()
		defer c.server.vidstreamMu.Unlock()

		// Close existing engine if any
		if c.server.vidstreamEngine != nil {
			c.server.vidstreamEngine.Close()
			c.server.vidstreamEngine = nil
		}

		// Create new engine with config
		config := vidstream.Config{
			ModelPath:           msg.ModelPath,
			ConfidenceThreshold: msg.ConfidenceThreshold,
			NMSThreshold:        msg.NMSThreshold,
			InputWidth:          640, // Default YOLOv11 input size
			InputHeight:         480,
			NumThreads:          0,     // Auto-detect
			UseGPU:              false, // CPU inference for now
			Labels:              "",    // Use default COCO labels
		}

		engine, err := vidstream.NewVideoEngineWithConfig(config)
		if err != nil {
			c.server.logger.Errorw("VidStream init failed",
				"error", err,
				"client_id", c.id,
			)
			c.sendMsg <- map[string]interface{}{
				"type":  "vidstream_init_error",
				"error": err.Error(),
			}
			return
		}

		c.server.vidstreamEngine = engine
		width, height := engine.InputDimensions()
		ready := engine.IsReady()

		c.server.logger.Infow("VidStream engine initialized",
			"client_id", c.id,
			"input_dimensions", fmt.Sprintf("%dx%d", width, height),
			"ready", ready,
		)

		c.sendMsg <- map[string]interface{}{
			"type":   "vidstream_init_success",
			"width":  width,
			"height": height,
			"ready":  ready,
		}
	}()
}

// handleVidStreamFrame processes a video frame through ONNX inference
func (c *Client) handleVidStreamFrame(msg QueryMessage) {
	c.server.vidstreamMu.Lock()
	defer c.server.vidstreamMu.Unlock()

	if c.server.vidstreamEngine == nil {
		c.sendMsg <- map[string]interface{}{
			"type":  "vidstream_frame_error",
			"error": "Engine not initialized. Call vidstream_init first.",
		}
		return
	}

	// Parse format
	var format vidstream.FrameFormat
	switch msg.Format {
	case "rgba8":
		format = vidstream.FormatRGBA8
	case "rgb8":
		format = vidstream.FormatRGB8
	default:
		c.sendMsg <- map[string]interface{}{
			"type":  "vidstream_frame_error",
			"error": fmt.Sprintf("Unsupported format: %s", msg.Format),
		}
		return
	}

	// Process frame
	result, err := c.server.vidstreamEngine.ProcessFrame(
		msg.FrameData,
		uint32(msg.Width),
		uint32(msg.Height),
		format,
		uint64(time.Now().UnixMicro()),
	)
	if err != nil {
		c.server.logger.Warnw("VidStream frame processing failed",
			"error", err,
			"client_id", c.id,
		)
		c.sendMsg <- map[string]interface{}{
			"type":  "vidstream_frame_error",
			"error": err.Error(),
		}
		return
	}

	// Send detections back to client
	c.sendMsg <- map[string]interface{}{
		"type":       "vidstream_detections",
		"detections": result.Detections,
		"stats": map[string]interface{}{
			"preprocess_us":    result.Stats.PreprocessUs,
			"inference_us":     result.Stats.InferenceUs,
			"postprocess_us":   result.Stats.PostprocessUs,
			"total_us":         result.Stats.TotalUs,
			"detections_raw":   result.Stats.DetectionsRaw,
			"detections_final": result.Stats.DetectionsFinal,
		},
	}
}

File: web/src-tauri/src/vidstream.rs
//! Video processing commands for Tauri desktop app
//!
//! Provides Tauri commands to interact with the vidstream inference engine.
//! Desktop-only - requires ONNX Runtime support.

use qntx_vidstream::types::VideoEngineConfig;
use qntx_vidstream::{FrameFormat, VideoEngine};
use serde::{Deserialize, Serialize};
use std::sync::Mutex;
use tauri::State;

/// Shared video engine state (thread-safe)
pub struct VideoEngineState {
    engine: Mutex<Option<VideoEngine>>,
}

impl VideoEngineState {
    pub fn new() -> Self {
        Self {
            engine: Mutex::new(None),
        }
    }
}

/// Configuration for initializing the video engine (from frontend)
#[derive(Debug, Deserialize)]
pub struct InitConfig {
    pub model_path: String,
    #[serde(default = "default_confidence")]
    pub confidence_threshold: f32,
    #[serde(default = "default_nms")]
    pub nms_threshold: f32,
    #[serde(default = "default_input_size")]
    pub input_width: u32,
    #[serde(default = "default_input_size")]
    pub input_height: u32,
}

fn default_confidence() -> f32 {
    0.5
}
fn default_nms() -> f32 {
    0.45
}
fn default_input_size() -> u32 {
    640
}

/// Detection result (serializable for Tauri)
#[derive(Debug, Serialize)]
pub struct DetectionResult {
    pub class_id: u32,
    pub label: String,
    pub confidence: f32,
    pub bbox: BBox,
    pub track_id: u64,
}

#[derive(Debug, Serialize)]
pub struct BBox {
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
}

/// Processing result with detections and stats
#[derive(Debug, Serialize)]
pub struct ProcessResult {
    pub detections: Vec<DetectionResult>,
    pub stats: Stats,
}

#[derive(Debug, Serialize)]
pub struct Stats {
    pub preprocess_us: u64,
    pub inference_us: u64,
    pub postprocess_us: u64,
    pub total_us: u64,
    pub detections_raw: u32,
    pub detections_final: u32,
}

/// Engine info (dimensions, ready status)
#[derive(Debug, Serialize)]
pub struct EngineInfo {
    pub ready: bool,
    pub input_width: u32,
    pub input_height: u32,
}

/// Initialize the video engine with a model
#[tauri::command]
pub fn vidstream_init(
    config: InitConfig,
    state: State<VideoEngineState>,
) -> Result<EngineInfo, String> {
    println!(
        "[vidstream] Initializing engine with model: {}",
        config.model_path
    );

    let engine_config = VideoEngineConfig {
        model_path: config.model_path,
        confidence_threshold: config.confidence_threshold,
        nms_threshold: config.nms_threshold,
        input_width: config.input_width,
        input_height: config.input_height,
        num_threads: 0, // auto
        use_gpu: false,
        labels: None,
    };

    let engine = VideoEngine::new(engine_config)?;
    let (width, height) = engine.input_dimensions();
    let ready = engine.is_ready();

    *state.engine.lock().unwrap() = Some(engine);

    println!(
        "[vidstream] Engine initialized: {}x{}, ready={}",
        width, height, ready
    );

    Ok(EngineInfo {
        ready,
        input_width: width,
        input_height: height,
    })
}

/// Process a video frame
#[tauri::command]
pub fn vidstream_process_frame(
    frame_data: Vec<u8>,
    width: u32,
    height: u32,
    format: String,
    timestamp_us: u64,
    state: State<VideoEngineState>,
) -> Result<ProcessResult, String> {
    let mut engine_lock = state.engine.lock().unwrap();
    let engine = engine_lock.as_mut().ok_or("Engine not initialized")?;

    // Parse format
    let frame_format = match format.as_str() {
        "rgb8" => FrameFormat::RGB8,
        "rgba8" => FrameFormat::RGBA8,
        "bgr8" => FrameFormat::BGR8,
        "yuv420" => FrameFormat::YUV420,
        "gray8" => FrameFormat::Gray8,
        _ => return Err(format!("Unknown format: {}", format)),
    };

    let (detections, stats) =
        engine.process_frame(&frame_data, width, height, frame_format, timestamp_us);

    Ok(ProcessResult {
        detections: detections
            .into_iter()
            .map(|d| DetectionResult {
                class_id: d.class_id,
                label: d.label,
                confidence: d.confidence,
                bbox: BBox {
                    x: d.bbox.x,
                    y: d.bbox.y,
                    width: d.bbox.width,
                    height: d.bbox.height,
                },
                track_id: d.track_id,
            })
            .collect(),
        stats: Stats {
            preprocess_us: stats.preprocess_us,
            inference_us: stats.inference_us,
            postprocess_us: stats.postprocess_us,
            total_us: stats.total_us,
            detections_raw: stats.detections_raw,
            detections_final: stats.detections_final,
        },
    })
}

/// Get current engine info
#[tauri::command]
pub fn vidstream_get_info(state: State<VideoEngineState>) -> Result<EngineInfo, String> {
    let engine_lock = state.engine.lock().unwrap();
    let engine = engine_lock.as_ref().ok_or("Engine not initialized")?;

    let (width, height) = engine.input_dimensions();
    let ready = engine.is_ready();

    Ok(EngineInfo {
        ready,
        input_width: width,
        input_height: height,
    })
}
